### ISSUE TITLE
Fails to infer correct value type

### ISSUE THREAD

**Issue by AndreasKostler** [CommBank/grimlock#45]
In swoop we ran across an issue where there are conflicting types for `DateValue` - `DateValue` neq `Value { type V = java.util.Date }`;

**Comment by poojke** [comment]
@AndreasKostler is this a grimlock issue, or a feature of the scala language?
```
scala> trait Foo { type X }
defined trait Foo

scala> case class Bar(i: Int) extends Foo { type X = Int }
defined class Bar

scala> def baz[F](foo: F)(implicit ev: F =:= Foo { type X = Int }) = foo.toString
baz: [F](foo: F)(implicit ev: =:=[F,Foo{type X = Int}])String

scala> baz(Bar(1))
<console>:20: error: Cannot prove that Bar =:= Foo{type X = Int}.
       baz(Bar(1))
          ^

scala> def baz[F](foo: F)(implicit ev: F <:< Foo { type X = Int }) = foo.toString
baz: [F](foo: F)(implicit ev: <:<[F,Foo{type X = Int}])String

scala> baz(Bar(1))
res11: String = Bar(1)
```
Based on the above it would suggest that the types shouldn't be the same. If so, are you happy for this issue to be closed?

**Comment by poojke** [comment]
Resolved with #57 - have the decode methods return the specific decoded `Value`.
